### ISSUE TITLE
Quickly Convert to HDF5 Format

### ISSUE THREAD

**Issue by ks233ever** [vaexio/vaex#812]
Hi Guys,

Able to reduce data processing time by about 25% using vaex -- bottleneck is the massive amount of time it takes to convert data to HDF5 format. Once in this format processing is instantaneous. Currently using vaex.from_csv to loop over the data in chunks, but still taking about 2 hrs to convert. Any ideas on how to quickly convert to HDF5 using vaex or another tool?

**Comment by maartenbreddels** [comment]
Hi Kayla,

thanks for posting your question here. From our private conversation, I understood that the original data source is from a database, is that correct? Doing a database->csv->hdf5 can be quite expensive I can imagine.

cheers,

Maarten

**Comment by alexrabe91** [comment]
but what would be a fast way to convert a csv file into hdf5 in order to enable the speed up with vaex? that actually looks like a real bottleneck for my/more generic use cases that are based on interacting with a database..
cheers!

**Comment by maartenbreddels** [comment]
Hi,

I think the database->csv->hdf5 not the best way to do this. I think database->arrow would be a much better way (and vaex can read arrow files).

https://turbodbc.readthedocs.io/en/latest/pages/advanced_usage.html#apache-arrow-support

I'd be really interested in finding out the performance increase this way, but I wouldn't be surprised if it can be 100x faster.

sth like:

```
cursor.execute("SELECT A, B FROM my_table")
batches = cursor.fetcharrowbatches()
for i, table in enumerate(batches):
  df_batch = vaex.from_arrow_table(table)
  df_batch.export(f'batch_{i:05}.arrow')
...
df = vaex.open('batch*.arrow')
df.export('single.hdf5')  # optional, it many small chunks, this might be faster to merge into 1 file, .arrow files should also be ok
```

Regards,

Maarten

**Comment by ks233ever** [comment]
Thanks for this! Will try this method and post the results.

**Comment by ks233ever** [comment]
Unfortunately I don't think this method will work either. Using a Redshift DB and the cursor.execute() to select 100 records from a 24 million table took a full minute to finish

**Comment by ks233ever** [comment]
I am trying to work with parquet files on s3 alternatively. Any idea why vaex.open('s3://bucket/folder/score_part_00.parquet') returns 

OSError: Passed non-file path: 's3://bucket/folder/score_part_00.parquet'

is this the correct syntax to use to open s3?

**Comment by JovanVeljanoski** [comment]
Hi @ks233ever 

For the moment only hdf5 files are streamable from S3 directly.

**Comment by ks233ever** [comment]
Can we call vaex.open() on file objects by any chance? 

For example running --

import fastparquet as fp

s3 = s3fs.S3FileSystem()
myopen = s3.open

fpfile = fp.ParquetFile('s3://XXXX', open_with=myopen)

df = vaex.open(fpfile)

returns error 

AttributeError: 'ParquetFile' object has no attribute 'startswith'

**Comment by ks233ever** [comment]
Noting here that I was able to reduce scoring time by 80% by using parquet with vaex. From 193 mins to 38 mins.

unload redshift to s3 in parquet format (creates 6 files) (15 mins)

download parquet files locally (5 mins)-- if s3 streaming for parquet becomes supported this could potentially go down further

call df = vaex.open() on each file

score each df (18 mins) -- I did try vaex.concat() on the list of df's in order to score once vs each idividual df. concat() ran for 30+ mins before I killed it

**Comment by maartenbreddels** [comment]
> From 193 mins to 38 mins.

That sounds good, thanks for letting us know?



> I did try vaex.concat() on the list of df's in order to score once vs each idividual df. concat() ran for 30+ mins before I killed it

I guess you have many columns? And your CPU usage is 100% (single core)? @JovanVeljanoski it's this issue again (which I *again* cannot find, can you find the reference for us?, I guess it's this known issue)

**Comment by JovanVeljanoski** [comment]
Yeah, I guess the concat performance is a known issue #604 (thread contains reproducible example). It is more prominent the more columns there are in the dataset.

**Comment by ks233ever** [comment]
That's right ~340 cols

**Comment by alexrabe91** [comment]
> Hi,
> 
> I think the database->csv->hdf5 not the best way to do this. I think database->arrow would be a much better way (and vaex can read arrow files).
> 
> https://turbodbc.readthedocs.io/en/latest/pages/advanced_usage.html#apache-arrow-support
> 
> I'd be really interested in finding out the performance increase this way, but I wouldn't be surprised if it can be 100x faster.
> 
> sth like:
> 
> ```
> cursor.execute("SELECT A, B FROM my_table")
> batches = cursor.fetcharrowbatches()
> for i, table in enumerate(batches):
>   df_batch = vaex.from_arrow_table(table)
>   df_batch.export(f'batch_{i:05}.arrow')
> ...
> df = vaex.open('batch*.arrow')
> df.export('single.hdf5')  # optional, it many small chunks, this might be faster to merge into 1 file, .arrow files should also be ok
> ```
> 
> Regards,
> 
> Maarten

Iam not able to combine the "turbodbc meets vaex" approach. Once I am trying to convert each batch containing an arrow table into a vaex DataFrame, iam getting the following error:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-28-43d42b3add9b> in <module>
      1 for i, table in enumerate(batches):
----> 2     df_v = vaex.from_arrow_table(table)

~/anaconda3/envs/vaex/lib/python3.6/site-packages/vaex/__init__.py in from_arrow_table(table)
    353     """
    354     from vaex_arrow.convert import vaex_df_from_arrow_table
--> 355     return vaex_df_from_arrow_table(table=table)
    356 
    357 def from_scalars(**kwargs):

~/anaconda3/envs/vaex/lib/python3.6/site-packages/vaex_arrow/convert.py in vaex_df_from_arrow_table(table)
    101 def vaex_df_from_arrow_table(table):
    102     from .dataset import DatasetArrow
--> 103     return DatasetArrow(table=table)

~/anaconda3/envs/vaex/lib/python3.6/site-packages/vaex_arrow/dataset.py in __init__(self, filename, table, write)
     20             self._load()
     21         else:
---> 22             self._load_table(table)
     23 
     24     def _load(self):

~/anaconda3/envs/vaex/lib/python3.6/site-packages/vaex_arrow/dataset.py in _load_table(self, table)
     49                 self._categories[name] = dict(labels=labels, N=len(labels))
     50             else:
---> 51                 column = column_from_arrow_array(arrow_array)
     52             self.columns[name] = column
     53             self.column_names.append(name)

~/anaconda3/envs/vaex/lib/python3.6/site-packages/vaex_arrow/convert.py in column_from_arrow_array(arrow_array)
     31     buffers = arrow_array.buffers()
     32     if len(buffers) == 2:
---> 33         return numpy_array_from_arrow_array(arrow_array)
     34     elif len(buffers) == 3 and  isinstance(arrow_array.type, type(pyarrow.string())):
     35         bitmap_buffer, offsets, string_bytes = arrow_array.buffers()

~/anaconda3/envs/vaex/lib/python3.6/site-packages/vaex_arrow/convert.py in numpy_array_from_arrow_array(arrow_array)
     72         array = numpy_bool_from_arrow_bitmap(bitmap, len(arrow_array) + offset)[offset:]
     73     else:
---> 74         array = np.frombuffer(data_buffer, dtype, len(arrow_array) + offset)[offset:]
     75 
     76     if bitmap_buffer is not None:

ValueError: buffer is smaller than requested size
```
However, it works when I am directly trying to create a pandas DataFrame from batch.

My conda env is based on the following packages:
`# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
abseil-cpp                20200225.2           he1b5a44_0    conda-forge
aplus                     0.11.0                     py_1    conda-forge
arrow-cpp                 0.17.1          py36h1234567_6_cpu    conda-forge
astropy                   4.0.1.post1      py36h8c4c3a4_0    conda-forge
attrs                     19.3.0                     py_0    conda-forge
aws-sdk-cpp               1.7.164              hc831370_1    conda-forge
backcall                  0.2.0              pyh9f0ad1d_0    conda-forge
bleach                    3.1.5              pyh9f0ad1d_0    conda-forge
blosc                     1.19.0               he1b5a44_0    conda-forge
bokeh                     2.1.0            py36h9f0ad1d_0    conda-forge
boost-cpp                 1.72.0               h8e57a91_0    conda-forge
boto3                     1.14.7             pyh9f0ad1d_0    conda-forge
botocore                  1.17.7             pyh9f0ad1d_0    conda-forge
bqplot                    0.12.12            pyh9f0ad1d_0    conda-forge
branca                    0.3.1                      py_0    conda-forge
brotli                    1.0.7             he1b5a44_1002    conda-forge
brotlipy                  0.7.0           py36h8c4c3a4_1000    conda-forge
bzip2                     1.0.8                h516909a_2    conda-forge
c-ares                    1.15.0            h516909a_1001    conda-forge
ca-certificates           2020.6.20            hecda079_0    conda-forge
cachetools                4.1.0                      py_1    conda-forge
certifi                   2020.6.20        py36h9f0ad1d_0    conda-forge
cffi                      1.11.5                   py36_0    conda-forge
chardet                   3.0.4           py36h9f0ad1d_1006    conda-forge
charls                    2.1.0                he1b5a44_2    conda-forge
click                     7.1.2              pyh9f0ad1d_0    conda-forge
cloudpickle               1.4.1                      py_0    conda-forge
contextvars               2.4                        py_0    conda-forge
cryptography              2.9.2            py36h45558ae_0    conda-forge
curl                      7.69.1               h33f0ec9_0    conda-forge
cycler                    0.10.0                     py_2    conda-forge
cytoolz                   0.10.1           py36h516909a_0    conda-forge
dask                      2.19.0                     py_0    conda-forge
dask-core                 2.19.0                     py_0    conda-forge
decorator                 4.4.2                      py_0    conda-forge
defusedxml                0.6.0                      py_0    conda-forge
distributed               2.19.0           py36h9f0ad1d_0    conda-forge
docutils                  0.15.2                   py36_0    conda-forge
entrypoints               0.3             py36h9f0ad1d_1001    conda-forge
freetype                  2.10.2               he06d7ca_0    conda-forge
fsspec                    0.7.4                      py_0    conda-forge
future                    0.18.2           py36h9f0ad1d_1    conda-forge
gflags                    2.2.2             he1b5a44_1002    conda-forge
giflib                    5.2.1                h516909a_2    conda-forge
glog                      0.4.0                h49b9bf7_3    conda-forge
grpc-cpp                  1.29.1               h9ea6770_0    conda-forge
h5py                      2.10.0          nompi_py36h513d04c_102    conda-forge
hdbcli                    2.4.202                  pypi_0    pypi
hdf5                      1.10.5          nompi_h3c11f04_1104    conda-forge
heapdict                  1.0.1                      py_0    conda-forge
icu                       64.2                 he1b5a44_1    conda-forge
idna                      2.9                        py_1    conda-forge
imagecodecs               2020.5.30        py36hc75b63b_1    conda-forge
imageio                   2.8.0                      py_0    conda-forge
immutables                0.14             py36h8c4c3a4_0    conda-forge
importlib-metadata        1.6.1            py36h9f0ad1d_0    conda-forge
importlib_metadata        1.6.1                         0    conda-forge
ipydatawidgets            4.0.1                      py_0    conda-forge
ipykernel                 5.3.0            py36h95af2a2_0    conda-forge
ipyleaflet                0.13.0             pyh9f0ad1d_0    conda-forge
ipympl                    0.5.6              pyh9f0ad1d_1    conda-forge
ipyscales                 0.5.0              pyh9f0ad1d_0    conda-forge
ipython                   7.15.0           py36h9f0ad1d_0    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
ipyvolume                 0.6.0a6            pyh9f0ad1d_0    conda-forge
ipyvue                    1.3.2              pyh9f0ad1d_0    conda-forge
ipyvuetify                1.4.0              pyh8c360ce_1    conda-forge
ipywebrtc                 0.5.0                    py36_0    conda-forge
ipywidgets                7.5.1                      py_0    conda-forge
jedi                      0.17.1           py36h9f0ad1d_0    conda-forge
jinja2                    2.11.2             pyh9f0ad1d_0    conda-forge
jmespath                  0.10.0             pyh9f0ad1d_0    conda-forge
jpeg                      9d                   h516909a_0    conda-forge
jsonschema                3.2.0            py36h9f0ad1d_1    conda-forge
jupyter_client            6.1.3                      py_0    conda-forge
jupyter_core              4.6.3            py36h9f0ad1d_1    conda-forge
jxrlib                    1.1                  h516909a_2    conda-forge
kiwisolver                1.2.0            py36hdb11119_0    conda-forge
krb5                      1.17.1               h2fd8d38_0    conda-forge
lcms2                     2.11                 hbd6801e_0    conda-forge
ld_impl_linux-64          2.33.1               h53a641e_7  
libaec                    1.0.4                he1b5a44_1    conda-forge
libblas                   3.8.0               14_openblas    conda-forge
libcblas                  3.8.0               14_openblas    conda-forge
libcurl                   7.69.1               hf7181ac_0    conda-forge
libedit                   3.1.20191231         h7b6447c_0  
libevent                  2.1.10               h72c5cf5_0    conda-forge
libffi                    3.3                  he6710b0_1  
libgcc-ng                 9.1.0                hdf63c60_0  
libgfortran-ng            7.5.0                hdf63c60_6    conda-forge
libiconv                  1.15              h516909a_1006    conda-forge
liblapack                 3.8.0               14_openblas    conda-forge
libllvm8                  8.0.1                hc9558a2_0    conda-forge
libopenblas               0.3.7                h5ec1e0e_6    conda-forge
libpng                    1.6.37               hed695b0_1    conda-forge
libprotobuf               3.12.3               h8b12597_0    conda-forge
libsodium                 1.0.17               h516909a_0    conda-forge
libssh2                   1.9.0                hab1572f_2    conda-forge
libstdcxx-ng              9.1.0                hdf63c60_0  
libtiff                   4.1.0                hc7e4089_6    conda-forge
libwebp-base              1.1.0                h516909a_3    conda-forge
libzopfli                 1.0.3                he1b5a44_0    conda-forge
llvmlite                  0.32.1           py36hfa65bc7_0    conda-forge
locket                    0.2.0                      py_2    conda-forge
lz4-c                     1.9.2                he1b5a44_1    conda-forge
markupsafe                1.1.1            py36h8c4c3a4_1    conda-forge
matplotlib-base           3.2.2            py36hb8e4980_0    conda-forge
mistune                   0.8.4           py36h8c4c3a4_1001    conda-forge
msgpack-python            1.0.0            py36hdb11119_1    conda-forge
nbconvert                 5.6.1            py36h9f0ad1d_1    conda-forge
nbformat                  5.0.6                      py_0    conda-forge
ncurses                   6.2                  he6710b0_1  
nest-asyncio              1.3.3                      py_0    conda-forge
networkx                  2.4                        py_1    conda-forge
nly-sdk                   0.0.1                    pypi_0    pypi
notebook                  6.0.3            py36h9f0ad1d_0    conda-forge
numba                     0.49.1           py36h830a2c2_0    conda-forge
numpy                     1.17.5           py36h95a1406_0    conda-forge
olefile                   0.46                       py_0    conda-forge
openjpeg                  2.3.1                h981e76c_3    conda-forge
openssl                   1.1.1g               h516909a_0    conda-forge
packaging                 20.4               pyh9f0ad1d_0    conda-forge
pandas                    1.0.5            py36h830a2c2_0    conda-forge
pandoc                    2.9.2.1                       0    conda-forge
pandocfilters             1.4.2                      py_1    conda-forge
parquet-cpp               1.5.1                         2    conda-forge
parso                     0.7.0              pyh9f0ad1d_0    conda-forge
partd                     1.1.0                      py_0    conda-forge
pcre                      8.44                 he1b5a44_0    conda-forge
pexpect                   4.8.0            py36h9f0ad1d_1    conda-forge
pickleshare               0.7.5           py36h9f0ad1d_1001    conda-forge
pillow                    7.1.2            py36h8328e55_0    conda-forge
pip                       20.1.1                   py36_1  
progressbar2              3.51.3             pyh9f0ad1d_0    conda-forge
prometheus_client         0.8.0              pyh9f0ad1d_0    conda-forge
prompt-toolkit            3.0.5                      py_1    conda-forge
psutil                    5.7.0            py36h8c4c3a4_1    conda-forge
ptyprocess                0.6.0                   py_1001    conda-forge
pyarrow                   0.17.1          py36h1234567_6_cpu    conda-forge
pybind11                  2.5.0                    pypi_0    pypi
pycparser                 2.20                       py_0    conda-forge
pygments                  2.6.1                      py_0    conda-forge
pyopenssl                 19.1.0                     py_1    conda-forge
pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge
pyrsistent                0.16.0           py36h8c4c3a4_0    conda-forge
pysocks                   1.7.1            py36h9f0ad1d_1    conda-forge
python                    3.6.10               h7579374_2  
python-dateutil           2.8.1                      py_0    conda-forge
python-utils              2.4.0                      py_0    conda-forge
python_abi                3.6                     1_cp36m    conda-forge
pythreejs                 2.2.0              pyh8c360ce_0    conda-forge
pytz                      2020.1             pyh9f0ad1d_0    conda-forge
pywavelets                1.1.1            py36h785e9b2_1    conda-forge
pyyaml                    5.3.1            py36h8c4c3a4_0    conda-forge
pyzmq                     19.0.1           py36h9947dbf_0    conda-forge
re2                       2020.06.01           he1b5a44_0    conda-forge
readline                  8.0                  h7b6447c_0  
requests                  2.24.0             pyh9f0ad1d_0    conda-forge
s3fs                      0.2.2                      py_0    conda-forge
s3transfer                0.3.3            py36h9f0ad1d_1    conda-forge
scikit-image              0.17.2           py36h830a2c2_1    conda-forge
scipy                     1.4.1            py36h2d22cac_3    conda-forge
send2trash                1.5.0                      py_0    conda-forge
setuptools                47.3.0                   py36_0  
six                       1.15.0             pyh9f0ad1d_0    conda-forge
snappy                    1.1.8                he1b5a44_2    conda-forge
sortedcontainers          2.2.2              pyh9f0ad1d_0    conda-forge
sqlite                    3.32.2               h62c20be_0  
tabulate                  0.8.7              pyh9f0ad1d_0    conda-forge
tblib                     1.6.0                      py_0    conda-forge
terminado                 0.8.3            py36h9f0ad1d_1    conda-forge
testpath                  0.4.4                      py_0    conda-forge
thrift-cpp                0.13.0               h62aa4f2_2    conda-forge
tifffile                  2020.6.3                   py_1    conda-forge
tk                        8.6.10               hbc83047_0  
toolz                     0.10.0                     py_0    conda-forge
tornado                   6.0.4            py36h8c4c3a4_1    conda-forge
traitlets                 4.3.3            py36h9f0ad1d_1    conda-forge
traittypes                0.2.1                      py_1    conda-forge
turbodbc                  4.0.0            py36h026a0f8_1    conda-forge
typing_extensions         3.7.4.2                    py_0    conda-forge
unixodbc                  2.3.7             hd1b7508_1000    conda-forge
urllib3                   1.25.9                     py_0    conda-forge
vaex                      3.0.0              pyh9f0ad1d_0    conda-forge
vaex-arrow                0.5.1              pyh9f0ad1d_0    conda-forge
vaex-astro                0.7.0              pyh9f0ad1d_0    conda-forge
vaex-core                 2.0.3            py36h830a2c2_0    conda-forge
vaex-hdf5                 0.6.0              pyh9f0ad1d_0    conda-forge
vaex-jupyter              0.5.2              pyh9f0ad1d_0    conda-forge
vaex-ml                   0.9.0              pyh9f0ad1d_0    conda-forge
vaex-server               0.3.1              pyh9f0ad1d_0    conda-forge
vaex-viz                  0.4.0              pyh9f0ad1d_0    conda-forge
wcwidth                   0.2.4              pyh9f0ad1d_0    conda-forge
webencodings              0.5.1                      py_1    conda-forge
wheel                     0.34.2                   py36_0  
widgetsnbextension        3.5.1                    py36_0    conda-forge
xarray                    0.15.1                     py_0    conda-forge
xz                        5.2.5                h7b6447c_0  
yaml                      0.2.5                h516909a_0    conda-forge
zeromq                    4.3.2                he1b5a44_2    conda-forge
zict                      2.0.0                      py_0    conda-forge
zipp                      3.1.0                      py_0    conda-forge
zlib                      1.2.11               h7b6447c_3  
zstd                      1.4.4                h6597ccf_3    conda-forge`

Iam on ubuntu 19.10.

Cheers

**Comment by maartenbreddels** [comment]
We just merged #517 which includes better arrow support, are you willing to test this? You'd need to install it from source.
